packages/apollo-engine-reporting/src/extension.ts

```diff
@@ -146,41 +146,57 @@ export class EngineReportingExtension<TContext = any>
       }
     }
 
-    if (this.options.privateVariables !== true && o.variables) {
-      // Note: we explicitly do *not* include the details.rawQuery field. The
-      // Engine web app currently does nothing with this other than store it in
-      // the database and offer it up via its GraphQL API, and sending it means
-      // that using calculateSignature to hide sensitive data in the query
-      // string is ineffective.
-      this.trace.details = new Trace.Details();
-      Object.keys(o.variables).forEach(name => {
-        if (
-          this.options.privateVariables &&
-          Array.isArray(this.options.privateVariables) &&
-          // We assume that most users will have only a few private variables,
-          // or will just set privateVariables to true; we can change this
-          // linear-time operation if it causes real performance issues.
-          this.options.privateVariables.includes(name)
-        ) {
-          // Special case for private variables. Note that this is a different
-          // representation from a variable containing the empty string, as that
-          // will be sent as '""'.
-          this.trace.details!.variablesJson![name] = '';
-        } else {
-          try {
-            this.trace.details!.variablesJson![name] = JSON.stringify(
-              o.variables![name],
-            );
-          } catch (e) {
-            // This probably means that the value contains a circular reference,
-            // causing `JSON.stringify()` to throw a TypeError:
-            // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Issue_with_JSON.stringify()_when_serializing_circular_references
-            this.trace.details!.variablesJson![name] = JSON.stringify(
-              '[Unable to convert value to JSON]',
-            );
+    if (
+      (this.options.enforcePrivateVariables ||
+        this.options.privateVariables === null) &&
+      o.variables
+    ) {
+      // Default to the enforcePrivateVariables specs if privateVariables was not set.
+      this.trace.details = makeTraceDetails(
+        o.variables,
+        this.options.enforcePrivateVariables,
+        o.queryString,
+      );
+    } else {
```

Comment:
seems like we might be able to pass in `enforcePrivateVariables || privateVariables` into the `makeTraceDetails` to avoid the copied code?

Reply:
David asked/commented about something similar, I wrote a response below but copying it here: 

> I separated these mostly because I didn't want to change the old logic/behavior, where if `privateVariables === true` then none of the variable keys would be recorded in the trace details (and I assumed the old option was working as expected, so I didn't write test cases for it-- my bad!). I'll wrap the old into another function, and do a couple more simple test cases.